pyro/infer/reparam/transform.py

```diff
@@ -0,0 +1,42 @@
+
+class TransformReparam:
+    """
+    Arbitrary transform reparameterizer.
+
+    This can be used to reparameterize wrt an arbitrary bijective
+    :class:`~torch.distributions.transforms.Transform` object, and requires:
+
+    - for sample sites, the forward ``.__call__()`` method and the
+      ``.log_abs_det_jacobian()`` transform to be defined, as in [1]
+    - for observation sites, requires the backward ``.inv.__call__()`` method
+      and the ``.log_abs_det_jacobian()``.
+
+    [1] Hoffman, M. et al. (2019)
+        "NeuTra-lizing Bad Geometry in Hamiltonian Monte Carlo Using Neural Transport"
+        https://arxiv.org/abs/1903.03704
+
+    :param ~pyro.distributions.TorchDistribution base_dist: A base
+        distribution for the auxiliary latent variable.
+    :param ~torch.distributions.transforms.Transform transform: A bijective
+        transform defining forward and log abs det jacobian methods.
+    """
+    def __init__(self, base_dist, transform):
+        self.base_dist = base_dist
+        self.transform = transform
+
+    def __call__(self, name, fn, obs):
```

Comment:
The direction of x-z data dependency here depends on whether the site is observed or not, which might change due to another handler. This seems problematic for 'poutine.condition`, 'poutine.replay` etc.